### ISSUE TITLE
Messages should not contain python objects 

### ISSUE THREAD

**Issue by rusty-snake** [mesonbuild/meson#9501]
**Describe the bug**

Python objects in (error-)messages are more difficult to understand than meson objects:

```
DEPRECATION: Setting a configuration data value to <mesonbuild.interpreter.interpreterobjects.RunProcess object at 0x7ff5e9948a90> is invalid, and will fail at configure_file(). If you are using it just to store some values, please use a dict instead.
```

should be

```
DEPRECATION: Setting a configuration data value to runresult is invalid, and will fail at configure_file(). If you are using it just to store some values, please use a dict instead.
```

**To Reproduce**

`meson.build`:
```meson
project('test01')

today = run_command('echo', 'today')  
   
conf_data = configuration_data() 
conf_data.set('TODAY', today)
```

cmd:
```bash
rm -rf _builddir; meson setup _builddir && meson compile -C _builddir -v
```

**Expected behavior**

The name of a meson object (e.g. `Run result` or `runresult`) instead of the name and address of a python object.

In this specific example it would be nice to suggest to use `.stdout()`/`.stderr()`.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? native
* What operating system? Fedora 34 Workstation
* What Python version are you using? 3.9.7
* What `meson --version`? 0.59.1
* What `ninja --version`? 1.10.2

**Comment by eli-schwartz** [comment]
I think part of the problem is that if we knew it was a "run_command result, please use .stdout()" we could also use stdout for you (maybe -- it is a question whether to use that or to ask whether you meant stderr, but the latter feels unlikely/uncommon).

But we don't, really. All we know is that it isn't a string/integer as expected, we don't enumerate the badness or offer code correction suggestions tailored to the particular way you're incorrectly using it.

Alternative error messages include, for example:

> meson.build:6: DEPRECATION: Setting a configuration data value to <CustomTarget today@cus: ['/usr/bin/date']> is invalid, and will fail at configure_file(). If you are using it just to store some values, please use a dict instead.

In that case, CustomTarget has a `__repr__()` which changes the format of the error message and technically is a python object, but looks less like one.

**Comment by dcbaker** [comment]
Now that everything is a meson object and not a mixture of base objects and meson objects, it might make sense to add a method to the Interpreter Objects for formatting errors, things like `CustomTarget (command : "/use/bin/date")` or something.

@mensinda probably knows better whether such a thing is feasible.

**Comment by eli-schwartz** [comment]
Yeah, printing a better name is feasible at least. But truthfully I don't see a whole lot of difference between that repr and your suggestion, so maybe we could just add sane repr methods to various functions that don't yet have them?

**Comment by dcbaker** [comment]
Well, repr is supposed to return a value that `eval` can use to create a new instance that is equivalent to the class, so repr really isn't the right thing here

**Comment by eli-schwartz** [comment]
At least it's consistently the wrong thing, and consistently wrong everywhere we use it...

**Comment by bonzini** [comment]
This should be mostly fixed, for example:

```
meson.build:6:10: ERROR: configuration_data.set argument 2 was of type "RunProcess" but should have been one of: "str", "int", "bool"
```

There may be occasional cases where it's still there, such as

```
        if len(args) != 1 or not isinstance(args[0], str):
            sys.stderr.write(f'Unable to evaluate subdir({args}) in AstInterpreter --> Skipping\n')
            return         
```

But interpreter type checking is by far the most common case. If anybody wants to collect the few cases that still exist, it can be done in a tracker bug.
